crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs

```diff
@@ -304,3 +295,105 @@ pub(crate) fn filter_contained(edits: Vec<Edit>) -> Vec<Edit> {
     }
     filtered
 }
+
+#[derive(Copy, PartialEq, Clone)]
+enum State {
+    Literal,
+    AnnotatedFirst,
+    AnnotatedRest,
+    Other,
+}
+
+pub(crate) struct QuoteAnnotation<'a> {
+    state: Vec<State>,
+    stylist: &'a Stylist<'a>,
+    annotation: String,
+    final_quote_type: Quote,
+}
+
+impl<'a> QuoteAnnotation<'a> {
+    pub(crate) fn new(stylist: &'a Stylist<'a>) -> Self {
+        let final_quote_type = stylist.quote();
+        Self {
+            state: vec![],
+            stylist,
+            annotation: String::new(),
+            final_quote_type,
+        }
+    }
+}
+
+impl<'a> source_order::SourceOrderVisitor<'a> for QuoteAnnotation<'a> {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        let generator = Generator::from(self.stylist);
+        match expr {
+            Expr::Subscript(ast::ExprSubscript { value, slice, .. }) => {
+                if let Some(name) = value.as_name_expr() {
```

Comment:
Should we only look for `Name` node here? What about attribute nodes like `pd.DataFrame` or `typing.Optional[int]` which is more relevant in this context as it's a subscript expression.

Reply:
Correct. Is it enough if we rely on an attribute with value of "typing" and attr of "Optional" or "Literal", etc.?
Or should we fully resolved and check the qualified name in this case that the symbol is resolved to that class or not?

Reply:
I added another case for attributes.